In [9]:
!pip install torchtext==0.16.0

  Using cached torchtext-0.16.0-cp311-cp311-manylinux1_x86_64.whl.metadata (7.5 kB)
  Using cached torch-2.1.0-cp311-cp311-manylinux1_x86_64.whl.metadata (25 kB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_nccl_cu12-2.18.1-py3-none-manylinux1_x86_64.whl.metadata (1.8 kB)
  Using cached triton-2.1.0-0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (1.3 kB)
Using cached torchtext-0.16.0-cp311-cp311-manylinux1_x86_64.whl (2.0 MB)
Using cached torch-2.1.0-cp311-cp311-manylinux1_x86_64.whl (670.2 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.6/410.6 MB 1.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.1/14.1 MB 52.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 82.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.6/823.6 kB 55.3 MB/s eta 0:00:00
Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
   ━━━━━

In [1]:
!pip install torch==2.5.1 transformers==4.29.0

  Using cached torch-2.5.1-cp311-cp311-manylinux1_x86_64.whl.metadata (28 kB)
  Using cached nvidia_cuda_nvrtc_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_runtime_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_cupti_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cudnn_cu12-9.1.0.70-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cublas_cu12-12.4.5.8-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cufft_cu12-11.2.1.3-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_curand_cu12-10.3.5.147-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cusolver_cu12-11.6.1.9-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cusparse_cu12-12.3.1.170-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_nccl_cu12-2.21.5-py3-none

In [3]:
import collections

import datasets
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
import torch.optim as optim
# import torchtext
import tqdm
import transformers
import matplotlib.pyplot as plt

In [4]:
print(torch.__version__)

2.5.1+cu124


In [5]:
import datasets

train_data, test_data = datasets.load_dataset("imdb", split=["train", "test"])

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/7.81k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/21.0M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/20.5M [00:00<?, ?B/s]

unsupervised-00000-of-00001.parquet:   0%|          | 0.00/42.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

In [46]:
train_data


Dataset({
    features: ['text', 'label'],
    num_rows: 25000
})

In [6]:
transformer_name = "bert-base-uncased"

tokenizer = transformers.AutoTokenizer.from_pretrained(transformer_name)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/file_download.py:795: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

best-base-uncased. In this bert-base is a type of bert which is small model with 12 transformer layers(encoder layers), 768 hidden dimension per layer, 12 attention heads, approximately 110 million parameters.
uncased means whatever text you will put in the model that will be lowercased.
Here we have got the tokenizer layer for our model.


In [11]:
tokenizer.tokenize("hello world !")

['hello', 'world', '!']

In [14]:
tokenizer.encode("hello world")

[101, 7592, 2088, 102]

Here we are giving special Ids to all the words in our input text and then by creating one hot vector of this ids we will put them into word embedding to create more detailed and short vectors. the id 101 is for start of sentence [CLS] and the id 102 is for the end of the sequence [SEP].

In [15]:
tokenizer("hello world")

{'input_ids': [101, 7592, 2088, 102], 'token_type_ids': [0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1]}

Here the token_type_ids tells that is the input single sentence or not. In some tasks like question answering we have two sentence but in this case we have only one sentence so all valuse are 0 in the vector, if there would have been a second sentence then there would be values as 1 in vector.
If we have given padding to our input sentence then we have to tell the model that it has to ignore this for that we have to put 0 in the attention_mask vector, but in this we have no padding so all values are 0.

In [7]:
def example_and_token(example, tokenizer):
  ids = tokenizer(example["text"], truncation=True)["input_ids"]
  return {"ids" : ids}

In [8]:
train_data = train_data.map(example_and_token, fn_kwargs={"tokenizer": tokenizer})

test_data = test_data.map(example_and_token, fn_kwargs={"tokenizer" : tokenizer})

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

In [49]:
train_data[0]

{'text': 'I rented I AM CURIOUS-YELLOW from my video store because of all the controversy that surrounded it when it was first released in 1967. I also heard that at first it was seized by U.S. customs if it ever tried to enter this country, therefore being a fan of films considered "controversial" I really had to see this for myself.<br /><br />The plot is centered around a young Swedish drama student named Lena who wants to learn everything she can about life. In particular she wants to focus her attentions to making some sort of documentary on what the average Swede thought about certain political issues such as the Vietnam War and race issues in the United States. In between asking politicians and ordinary denizens of Stockholm about their opinions on politics, she has sex with her drama teacher, classmates, and married men.<br /><br />What kills me about I AM CURIOUS-YELLOW is that 40 years ago, this was considered pornographic. Really, the sex and nudity scenes are few and far be

In [9]:
test_size = 0.20

train_valid_data = train_data.train_test_split(test_size=test_size)
train_data = train_valid_data["train"]
val_data = train_valid_data["test"]

In [10]:
train_data = train_data.with_format(type="torch", columns=["ids", "label"])
val_data = val_data.with_format(type="torch", columns=["ids", "label"])
test_data = test_data.with_format(type="torch", columns=["ids", "label"])

In [11]:
def get_collate_fn(pad_index):
  def collate_fn(batch):
    batch_ids = [i["ids"] for i in batch]
    batch_ids = nn.utils.rnn.pad_sequence(batch_ids, padding_value=pad_index, batch_first=True)
    batch_label = [i["label"] for i in batch]
    batch_label = torch.stack(batch_label)
    batch = {"ids": batch_ids, "label" : batch_label}
    return batch

  return collate_fn

This function is to make all tokens of our training data of equal size and stack the batches together

In [12]:
pad_index = tokenizer.pad_token_id

In [13]:
def get_data_loader(dataset, batch_size, pad_index, shuffle=False):
  collate_fn = get_collate_fn(pad_index)
  data_loader = torch.utils.data.DataLoader(
      dataset = dataset,
      batch_size = batch_size,
      collate_fn = collate_fn,
      shuffle = shuffle,
  )
  return data_loader

In [14]:
batch_size = 8

train_data_loader = get_data_loader(train_data, batch_size, pad_index, shuffle=True)
val_data_loader = get_data_loader(val_data, batch_size, pad_index)
test_data_loader = get_data_loader(test_data, batch_size, pad_index)

In [15]:
class Transformer(nn.Module):
  def __init__(self, transformer, output_dim, freeze):
    super().__init__()
    self.transformer = transformer
    hidden_dim = transformer.config.hidden_size
    self.fc = nn.Linear(hidden_dim, output_dim)
    if freeze: 
      for param in self.transformer.parameters():
        param.requires_grad = False

  def forward(self, ids):
    # ids = [batch_size, seq_len]
    output = self.transformer(ids, output_attentions=True)
    hidden = output.last_hidden_state
    # hidden = [batch_size, seq_len, hidden_dim]
    attention = output.attentions[-1]
    # attention = [batch_size, n_heads, seq_len, seq_len]
    cls_hidden = hidden[:, 0, :]
    prediction = self.fc(torch.tanh(cls_hidden))
    # prediction = [batch_size, output_dim]
    return prediction

In [16]:
transformer = transformers.AutoModel.from_pretrained(transformer_name)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/file_download.py:795: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.seq_relationship.weight', 'cls.predictions.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.LayerNorm.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [17]:
transformer.config.hidden_size

768

In [18]:
output_dim = len(train_data['label'].unique())
freeze = False

model = Transformer(transformer, output_dim, freeze)

In [19]:
def count_parameters(model):
  return sum(p.numel() for p in model.parameters() if p.requires_grad)

print(f"the model has {count_parameters(model):,} trainable parameters")

the model has 109,483,778 trainable parameters


In [20]:
lr = 1e-6

optimizer = optim.Adam(model.parameters(), lr=lr)

In [21]:
criterion = nn.CrossEntropyLoss()

In [22]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
device

device(type='cuda')

In [23]:
model = model.to(device)
criterion = criterion.to(device)

In [24]:
def train(data_loader, model, criterion, optimizer, device):
  model.train()
  epoch_losses = []
  epoch_accs = []
  for batch in tqdm.tqdm(data_loader, desc="training..."):
    ids = batch["ids"].to(device)
    label = batch["label"].to(device)
    prediction = model(ids)
    loss = criterion(prediction, label)
    accuracy = get_accuracy(prediction, label)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    epoch_losses.append(loss.item())
    epoch_accs.append(accuracy.item())
  return np.mean(epoch_losses), np.mean(epoch_accs)

In [28]:
def evaluate(data_loader, model, criterion, device):
  model.eval()
  epoch_losses = []
  epoch_accs = []
  with torch.no_grad():
    for batch in tqdm.tqdm(data_loader, desc="evaluating..."):
      ids = batch["ids"].to(device)
      label = batch["label"].to(device)
      prediction = model(ids)
      loss = criterion(prediction, label)
      accuracy = get_accuracy(prediction, label)
      epoch_losses.append(loss.item())
      epoch_accs.append(accuracy.item())
  return np.mean(epoch_losses), np.mean(epoch_accs)

In [26]:
def get_accuracy(prediction, label):
  batch_size, _ = prediction.shape
  predicted_classes = prediction.argmax(dim=-1)
  correct_predictions = predicted_classes.eq(label).sum()
  accuracy = correct_predictions / batch_size
  return accuracy

In [30]:
n_epochs = 3
best_valid_loss = float("inf")

metrics = collections.defaultdict(list)

for epoch in range(n_epochs):
  train_loss, train_acc = train(train_data_loader, model, criterion, optimizer, device)
  valid_loss, valid_acc = evaluate(val_data_loader, model, criterion, device)

  metrics["train_losses"].append(train_loss)
  metrics["train_accs"].append(train_acc)
  metrics["valid_losses"].append(valid_loss)
  metrics["valid_acc"].append(valid_acc)

  if valid_loss < best_valid_loss:
    best_valid_loss = valid_loss
    torch.save(model.state_dict(), "transformer.pt")

  print(f"epoch: {epoch}")
  print(f"train_loss: {train_loss:.3f}, train_acc: {train_acc:.3f}")
  print(f"valid_loss: {valid_loss:.3f}, valid_acc: {valid_acc:.3f}")

evaluating...: 100%|██████████| 625/625 [00:35<00:00, 17.56it/s]


epoch: 0
train_loss: 0.210, train_acc: 0.917
valid_loss: 0.222, valid_acc: 0.913


evaluating...: 100%|██████████| 625/625 [00:35<00:00, 17.57it/s]


epoch: 1
train_loss: 0.182, train_acc: 0.932
valid_loss: 0.220, valid_acc: 0.916


evaluating...: 100%|██████████| 625/625 [00:35<00:00, 17.47it/s]

epoch: 2
train_loss: 0.160, train_acc: 0.942
valid_loss: 0.227, valid_acc: 0.919


In [31]:
model.load_state_dict(torch.load("transformer.pt"))

test_loss, test_acc = evaluate(test_data_loader, model, criterion, device)

<ipython-input-31-657863e4a791>:1: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  model.load_state_dict(torch.load("transformer.pt"))
evaluating...: 100%|██████████| 3125/312

In [32]:
print(f"test_loss : {test_loss:.3f}, test_acc : {test_acc:.3f}")

test_loss : 0.192, test_acc : 0.926


In [34]:
def predict_sentiment(text, model, tokenizer, device):
  ids = tokenizer(text)["input_ids"]
  tensor = torch.LongTensor(ids).unsqueeze(dim=0).to(device)
  prediction = model(tensor).squeeze(dim=0)
  probability = torch.softmax(prediction, dim=-1)
  predicted_class = prediction.argmax(dim=-1).item()
  predicted_probability = probability[predicted_class].item()
  return predicted_class, predicted_probability

In [38]:
text = "This film is not terrible, it's great!"

predict_sentiment(text, model, tokenizer, device)

(1, 0.9453592300415039)